### ISSUE TITLE
image() calls putImageData() even if the imageData hasn't changed

### ISSUE THREAD

**Issue by piersh** [processing-js/processing-js#218]
````
var gr = createGraphics(400,400,P2D);
image(gr, 0, 0);
image(gr, 0, 0);
````

notice that during the 2nd `image()` call, the contents of `gr` hasn't changed, but `getCanvasData()` initializes a canvas and calls `putImageData()` again anyway. for repeated drawing of large images this is incredibly wasteful. it would be much better to have `gr` cache its canvas until the next `gr.beginDraw()` call.

on my machine here, each 400x400 image call takes 123ms (95% of which is in `putImageData()`), so drawing just 8 images drops my FPS down to 1. per second.

**Comment by Pomax** [comment]
the core question for this is "how do you know gt is the same data", when you're processing.js: in JS, there is no stability guarantee, so two consecutive calls like the one you show can actually have different data in `gr` simply by virtue of having accessed it. Pjs can't afford to check all the input, it has to trust your code to be reasonably efficient already, because it is simply going to translate your code from Processing syntax to javascript syntax, then run it, so the first step (annoying as it might sound) is to first make sure your own code, regardless of Process vs. Pjs, handles data intelligently. If an image hasn't changed, don't call `image(...)` a second time in a row.

**Comment by piersh** [comment]
i don't think you understood what i was trying to say. the problem isn't with code transpilation, it's with the implementation of the Processing.image() call, specifically that it doesn't cache the generated context between beginDraw() calls resulting in a significant performance degradation drawing multiple instances of the same createGraphics-created sprite.

> how do you know gt is the same data
you know the data is the same if `gr.beginDraw()` hasn't been called. you don't need to "check all the input", just can just invalidate the cached canvas on beginDraw().

> If an image hasn't changed, don't call image(...) a second time in a row.
ok, consider this code:
```
image(gr, 0, 0);
image(gr, 0, 1);
```
obviously the second call is required. my initial example was pathological, but the problem still exists: consider code that draws hundred of the same sprite in different locations. why is it necessary to call `putImageData()` for each draw?

**Comment by GoToLoop** [comment]
> BTW: not sure I appreciate the insta-close without any discussion.

Unfortunately I haven't found any place across GitHub where such behavior doesn't exist!  =(

> ... that it doesn't cache the generated context between **beginDraw()** calls...

Processing Java Mode caches **image()** calls in 1 _**WeakHashMap**_ container:
http://docs.Oracle.com/javase/8/docs/api/java/util/WeakHashMap.html

Functions **set()** & **background()** aren't cached though. So there's a way out to avoid auto-cache.

JS got _**WeakMap**_ instead. Though ES6, it exists for a long time now. All major browsers support it:
https://developer.Mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakMap

However for PJS project to rely on that it needs to abandon its insistent stance to try to keep compatibility w/ dinosaurs like IE10, Safari 6, etc.

**Comment by Pomax** [comment]
the instaclose was me hitting close instead of comment while rushing between sessions at an all-staff workweek, let's reopen that.

**Comment by piersh** [comment]
ok, i have found a couple of workarounds just in case anyone else is running into this.

the first is to generate an image from the 'graphics' and then use that:
````
var gr = createGraphics(400,400,P2D);
var img = gr.get();
image(img, 0, 0);
image(img, 0, 0);
````
however, this `get()` call still persists `gr` out to an array and then creates a whole new canvas and copies the array back into the image. not good for off-screen sprite rendering.

the alternative is this:
````
var gr = createGraphics(400,400,P2D);

var img = createImage(gr.width, gr.height, RGB);
img.sourceImg = gr.externals.canvas;
img.remote = true;

image(img, 0, 0);
image(img, 0, 0);
````
this allows the `image()` call to blit directly from gr's canvas into the main canvas without persiting the image out to an array and constructing a new image from the array *every* draw call.

i'm not familiar with every drawing feature going on here, but it would seem to me that the default drawing path should be this. sure, if you need to do special stuff with tints, etc... then fall-back to the slow path, otherwise just use the canvas API directly.
